e2e/cypress/integration/messaging/message_reaction_gm_spec.js

```diff
@@ -0,0 +1,85 @@
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Group: @messaging
+
+describe('Emoji reactions to posts/messages in GM channels', () => {
+    let userOne;
+    let userTwo;
+    let testTeam;
+    let testGroupChannel;
+
+    before(() => {
+        cy.apiInitSetup().then(({team, user}) => {
+            testTeam = team;
+            userOne = user;
+
+            cy.apiCreateUser().then((data) => {
+                userTwo = data.user;
+
+                cy.apiAddUserToTeam(testTeam.id, userTwo.id);
+
+                cy.apiCreateGroupChannel([userOne.id, userTwo.id]).then((response) => {
+                    testGroupChannel = response.body;
+                });
+
+                // # Login as userOne and town-square
+                cy.apiLogin(userOne);
+                cy.visit(`/${testTeam.name}/channels/town-square`);
+            });
+        });
+    });
+
+    it('MM-T471 add a reaction to a message in a GM', () => {
+        // # Switch to the GM
+        cy.visit(`/${testTeam.name}/messages/${testGroupChannel.name}`);
+
+        // # Post a message
+        cy.postMessage('This is a post');
+
+        cy.getLastPostId().then((postId) => {
+            // # Click the add reaction icon
+            cy.clickPostReactionIcon(postId);
+
+            // # Choose "slightly_frowning_face" emoji
+            // delaying 500ms in case of lag
+            cy.get('.emoji-picker__items #emoji-1f641').wait(500).click();
+
+            // * The number shown on the reaction is incremented by 1
+            cy.get(`#postReaction-${postId}-slightly_frowning_face .Reaction__number--display`).
+                should('have.text', '1').
+                should('be.visible');
+        });
+    });
+
+    it('MM-T471 the Add a Reaction button should only be visible when hovering the post or when in mobile vieew', () => {
+        cy.getLastPostId().then((postId) => {
+            // * Verify that the Add Reaction button isn't visible
+            cy.findByLabelText('Add a reaction').should('not.be.visible');
+
+            // # Focus on the post since we can't hover with Cypress
+            cy.get(`#post_${postId}`).focus().tab().tab();
+
+            // * Verify that the Add Reaction button is 1now visible
+            cy.findByLabelText('Add a reaction').should('be.visible');
+
+            // # Click somewhere to clear the focus
+            cy.get('#channelIntro').click();
+
+            // * Verify that the Add Reaction button is no longer visible
+            cy.findByLabelText('Add a reaction').should('not.be.visible');
+
+            // # Resize window to mobile view
+            cy.viewport('iphone-6');
+
+            // * Verify that the Add Reaction button is once again visible
+            cy.findByLabelText('Add a reaction').should('be.visible');
```

Comment:
@hmhealey We had some discussions about adopting written scripts against the original test case in TM4J, especially on how to break the test into multiple steps or extract into its own test case. Just like to chime in here before updating the TM4J. Would it be possible to combine above steps in the first `it` block? If no, can we run the second `it` block in isolation by running `it.only`?  Thanks

Reply:
Yeah, I can combine the two, but would it be better to maybe repeat the setup steps for the second one and keep them isolated? I was thinking of the two as covering different pieces of functionality, but the second one certainly won't run without the first. I'm not sure when we prefer to reuse test setup to save time and when we want the tests to remain isolated since I think we tend towards reuse normally

Reply:
Yeah, I think it's fine and I agree on keeping test isolated. I'll go on test isolation first then reusability. It makes it easier to deal with during test maintenance.  Thanks Harrison!

Reply:
I ended up just combining the test cases for now since I've been focused on some other stuff. When I get back to unit tests, I might look at splitting the tests up again.